### PR TITLE
[Robot] Validate hierarchy settings and backfill settings

### PR DESCRIPTION
**Pull request by rhakeshj**
Verify backfill settings when course connections is enabled
Verify org wide custom settings are not over written by user custom settings

Robot WI:
W-8448330
W-8448341

New/Existing Test - This is a new test

To run the test case, run this from your CCI:
cci task run robot -o suites robot/EDA/tests/browser/settings/hierarchy_settings.robot
cci task run robot -o suites robot/EDA/tests/browser/settings/course_connections_backfill.robot

# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

# Testing Notes

### CODE REVIEW

File: robot/EDA/resources/CourseConnectionsSettingsPageObject.py [hunk 1]
```diff
@@ -42,6 +49,20 @@ def set_enable_course_connections(self):
                 return
         raise Exception("Clicking element 'Enable Course Connections' failed after multiple tries")
 
+    def select_backfill_checkbox(self):
```
Comment (skristem) [inline robot/EDA/resources/CourseConnectionsSettingsPageObject.py:52]: might be a good idea to check first the status of the checkbox. If it is already checked then u do not want to change the status. I am thinking of scenarios where the test case with this keyword gets executed multiple times in the same org.

File: robot/EDA/resources/SetupPageObject.py [hunk 2]
```diff
@@ -13,10 +15,34 @@ def _go_to_page(self):
...
     def perform_a_quick_find_on(self, item):
         """ Searches for the given item through Quick Find on Setup page """
         self.eda.populate_placeholder("Quick Find", item)
 
+    def select_frame(self, value):
+        """ Selects iframe using the value passed from the test. The value can be iframe id, title
+            etc
+        """
+        locator = eda_lex_locators["eda_setup"]["frame"].format(value)
+        time.sleep(4)
```
Comment (skristem) [inline robot/EDA/resources/SetupPageObject.py:43]: can u wait for the frame instead of just sleep here

File: robot/EDA/resources/locators_50.py [hunk 3]
```diff
@@ -25,6 +25,12 @@
         "tab": "//div[@class='uiTabBar']/ul[@class='tabs__nav']/li[contains(@class,'uiTabItem')]/a[@class='tabHeader']/span[contains(text(), '{}')]",
         "spl-tab": "//div[@class='slds-tabs_default']//ul[@class='slds-tabs_default__nav']/li[contains(@class,'slds-tabs_default__item')]/a[text()= '{}']",
     },
+    "eda_setup": {
+        "custom_settings": "//a[text()='{}']",
```
Comment (skristem) [inline robot/EDA/resources/locators_50.py:29]: since this is a very common locator, you might want to move it to the outer dictionary. Just a suggestion

File: robot/EDA/resources/CourseConnectionsSettingsPageObject.py (hunk 1 above)
Comment (rhakeshj) [reply robot/EDA/resources/CourseConnectionsSettingsPageObject.py:52]: Yeah I thought about it. But this case is unique here. In this page, this checkbox enables the button to initiate a batch/queue kinda process. Once that is done and if the user moves out of the screen, the check box automatically flips to unchecked. So for every iteration this checkbox will load as unchecked avoiding the necessary to check for it

File: robot/EDA/resources/CourseConnectionsSettingsPageObject.py [hunk 4]
```diff
@@ -42,6 +49,20 @@ def set_enable_course_connections(self):
...
+        """ Selects the backfill checkbox """
+        locator = eda_lex_locators["eda_settings_cc"]["backfill_checkbox"]
+        self.selenium.wait_until_page_contains_element(locator)
+
+        for i in range(3):
+            i += 1
```
Comment (Safiyab) [inline robot/EDA/resources/CourseConnectionsSettingsPageObject.py:58]: You don't have to increment, the for loop will take care of it

File: robot/EDA/resources/SetupPageObject.py [hunk 5]
```diff
@@ -13,10 +15,34 @@ def _go_to_page(self):
...
+        locator = eda_lex_locators["eda_setup"]["settings_action_button"].format(button)
+        self.selenium.wait_until_page_contains_element(locator)
+        self.selenium.wait_until_element_is_visible(locator)
+        self.selenium.click_element(locator)
+
     def perform_a_quick_find_on(self, item):
         """ Searches for the given item through Quick Find on Setup page """
         self.eda.populate_placeholder("Quick Find", item)
 
+    def select_frame(self, value):
```
Comment (Safiyab) [inline robot/EDA/resources/SetupPageObject.py:38]: This is a more generic keyword. I think you should move it to EDA.py

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    Select accounts contacts checkbox
+    ...         Account Types with Multi-Addresses Enabled=Household Account
+    Click action button on EDA settings page    Save
+    Load Page Object                     Home                          Setup
+    Go to custom settings setup
+    Select frame        Custom Settings ~ Salesforce - Developer Edition
+    Click custom settings       Hierarchy Settings
+    Select frame        Custom Setting Definition ~ Salesforce - Developer Edition
+    Click custom settings action button     Manage
+    Select frame        Custom Setting Hierarchy Settings ~ Salesforce - Developer Edition
```
Comment (Safiyab) [inline robot/EDA/tests/browser/settings/hierarchy_settings.robot:27]: I think this all can be one keyword like: `Go to hierarchy settings page`
You don't need to show this level of detail in a test for a setup keyword. Especially the select and unselect frame needs to be hidden in the implementation.

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    Verify setup owner section      No records to display.
+    Unselect frame
+    Go to custom settings setup
```
Comment (Safiyab) [inline robot/EDA/tests/browser/settings/hierarchy_settings.robot:30]: Why is this the last step in the file ?

**Review by Safiyab — APPROVED** [review]
Made some suggestions. Can you please take a look and do the updates (they are nice-to have but not necessarily blocking)

File: robot/EDA/resources/CourseConnectionsSettingsPageObject.py (hunk 4 above)
Comment (rhakeshj) [reply robot/EDA/resources/CourseConnectionsSettingsPageObject.py:58]: That makes the iterator variable unused (other than we can use it for logging the iteration). We could use while for retries but again it will go undefined limits of attempt until the condition passes

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot (hunk 6 above)
Comment (rhakeshj) [reply robot/EDA/tests/browser/settings/hierarchy_settings.robot:27]: Agreed. I will move these under one keyword.

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot (hunk 7 above)
Comment (rhakeshj) [reply robot/EDA/tests/browser/settings/hierarchy_settings.robot:30]: This is added to make sure it leaves out of the frames (like switch to default frame)

File: robot/EDA/resources/SetupPageObject.py (hunk 2 above)
Comment (rhakeshj) [reply robot/EDA/resources/SetupPageObject.py]: @skristem This has been updated and moved to EDA.py as suggested

File: robot/EDA/resources/locators_50.py (hunk 3 above)
Comment (rhakeshj) [reply robot/EDA/resources/locators_50.py:29]: Already had a similar locator in eda dictionary and made use of it and this is removed now

File: robot/EDA/resources/CourseConnectionsSettingsPageObject.py (hunk 4 above)
Comment (rhakeshj) [reply robot/EDA/resources/CourseConnectionsSettingsPageObject.py]: Removed the iterator increment and replaced with logging

File: robot/EDA/resources/SetupPageObject.py (hunk 5 above)
Comment (rhakeshj) [reply robot/EDA/resources/SetupPageObject.py]: Done

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot (hunk 6 above)
Comment (rhakeshj) [reply robot/EDA/tests/browser/settings/hierarchy_settings.robot]: Moved to a common keyword now

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot (hunk 7 above)
Comment (rhakeshj) [reply robot/EDA/tests/browser/settings/hierarchy_settings.robot]: Moved this step to Test Teardown

File: robot/EDA/resources/locators_50.py (hunk 3 above)
Comment (skristem) [reply robot/EDA/resources/locators_50.py:29]: I still see the locator

File: robot/EDA/tests/browser/settings/course_connections_backfill.robot [hunk 8]
```diff
@@ -7,20 +7,39 @@ Library         cumulusci.robotframework.PageObjects
...
+    Click action button on EDA settings page    Save
+    Select course connections subtab            Backfill
+    Select backfill checkbox
+    Click backfill button
+    Verify backfill toast message       The backfill was queued successfully.
+    Reload Page
+    Go to EDA settings tab          Course Connections
+    Select course connections subtab            Backfill
+    Verify backfill checkbox value              false
+
```
Comment (skristem) [inline robot/EDA/tests/browser/settings/course_connections_backfill.robot:42]: I suggest adjusting the formatting for easy readability in both this and below test

**Review by skristem — APPROVED** [review]
minor suggestions looks good otherwise

File: robot/EDA/tests/browser/settings/hierarchy_settings.robot [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    ...         Account Types with Multi-Addresses Enabled=Household Account
+    Click action button on EDA settings page    Save
+    Load Page Object                     Home                          Setup
+    Go to custom settings setup
+    Select hierarchy settings        Custom Settings ~ Salesforce - Developer Edition
+    ...                             Hierarchy Settings
+    Wait until loading is complete
+    Go to hierarchy settings view       Custom Setting Definition ~ Salesforce - Developer Edition
+    ...                                 Manage
+    Wait until loading is complete
```
Comment (Safiyab) [inline robot/EDA/tests/browser/settings/hierarchy_settings.robot:30]: I think you can hide the frame value and the waits too. This is what I had in mind:
```
Go to custom settings in setup
Select hierarchy settings           mode=Manage
```

You can play around with the level of details here. But leave this for another day and another PR.
This test looks and works great.

**Review by Safiyab — APPROVED** [review]

**Review by dgamboni — APPROVED** [review]
Functional review looks great, thank you!

**Review by spinkelman — APPROVED** [review]
After Functional Review, This item is approved.
